R/adducts.R

```diff
@@ -0,0 +1,449 @@
+                        .adductRules(polarity = "negative"))
+
+  # check if adduct suppplied is in the list of valid adducts
+  if(!adduct %in% names(adduct_rules_all)) {
+    
+    stop(paste0("Unknown adduct: ", adduct))
+    
+  }
+  
+  # retrieve multiplicative and additive part for calculation
+  mass_multi <- adduct_rules_all[[adduct]]$mass_multi
+  mass_add <- adduct_rules_all[[adduct]]$mass_add
+  
+  ion_mass <- x * mass_multi + mass_add
+  
+  ion_mass
+  
+}
+
+#' @title Calculate neutral mass
+#'
+#' @description
+#'
+#' `mz2mass` calculates the neutral mass from a given m/z value and adduct 
+#'      definition
+#'
+#' @param x `numeric` m/z value for which the neutral mass shall be calculated.
+#'
+#' @param adduct `character` Valid adduct definition.
+#'
+#' @return `numeric` neutral mass calculated
+#'
+#' @author Michael Witting
+#'
+#' @export
+#'
+#' @examples
+#'
+#' ion_mass <- 100
+#' adduct <- "[M+H]+"
+#' 
+#' ## Calculate m/z of [M+H]+ adduct from neutral mass
+#' mz2mass(ion_mass, adduct)
+#' 
+#' ion_mass <- 100
+#' adduct <- "[M+Na]+"
+#' 
+#' ## Calculate m/z of [M+Na]+ adduct from neutral mass
+#' mz2mass(ion_mass, adduct)
+mz2mass <- function(x, adduct = "[M+H]+") {
+  
+  adduct_rules_all <- c(.adductRules(polarity = "positive"),
+                        .adductRules(polarity = "negative"))
+  
+  # check if adduct suppplied is in the list of valid adducts
+  if(!adduct %in% names(adduct_rules_all)) {
+    
+    stop(paste0("Unknown adduct: ", adduct))
+    
+  }
+  
+  # retrieve multiplicative and additive part for calculation
+  mass_multi <- adduct_rules_all[[adduct]]$mass_multi
+  mass_add <- adduct_rules_all[[adduct]]$mass_add
+  
+  exact_mass <- (x - mass_add) / mass_multi
+  
+  exact_mass
+  
+}
+
+#' @title Retrieve names of adducts
+#'
+#' @description
+#'
+#' `adductNames` allows to retrieve all valid adduct names currently supported
+#'
+#' @param polarity `character` Definition of ion mode, either "positive" or "negative"
+#'
+#' @return `character` A vector of all valid adduct names for the selected ion
+#'     mode
+#'
+#' @author Michael Witting
+#'
+#' @export
+#'
+#' @examples
+#'
+#' ## retrieve names of adduct names in positive ion mode
+#' adductNames(polarity = "positive")
+#' 
+#' ## retrieve names of adduct names in negative ion mode
+#' adductNames(polarity = "negative)
+#' 
+adductNames <- function(polarity = c("positive", "negative")) {
+  
+  if(polarity == "positive") {
+    
+    return(names(.adductRulesPos()))
+    
+  } else if(polarity == "negative") {
+    
+    return(names(.adductRulesNeg()))
+    
+  } else {
+    
+    stop("Unknown value for parameter polarity, use either 'positive' or 'negative'")
+    
+  }
+}
+
+
+#===============================================================================
+# Private functions
+#===============================================================================
+.adductRules <- function(polarity = c("positive", "negative")) {
+  
```

Comment:
I would add here a `match.arg(polarity)` - then you don't need the final `else` and the `stop` below.